book/variables_and_subexpressions.md

```diff
@@ -1,31 +1,133 @@
+| `=`         | Assigns a new value to the variable                                        |
+| `+=`        | Adds a value to the variable and makes the sum its new value               |
+| `-=`        | Subtracts a value from the variable and makes the difference its new value |
+| `*=`        | Multiplies the variable by a value and makes the product its new value     |
+| `/=`        | Divides the variable by a value and makes the quotient its new value       |
+| `**=`       | Concatenates the variable with a list making the new list its new value   |
+
+> **Note**
+> 1. `+=`, `-=`, `*=` and `/=` are only valid in the contexts where their root operations
+> are expected to work. For example, `+=` uses addition, so it can not be used for contexts
+> where addition would normally fail
+> 2. `**=` requires a variable representing a list **and** a list argument
+
+###### More on Mutability
+
+Closures and nested `def`s cannot capture mutable variables from their environment. For
+example
+
+```
+# naive method to count number of elements in a list
+mut x = 0
+
+[1 2 3] | each { $x += 1 }   # error: $x is captured in a closure
+```
+
+To use mutable variables for such behaviour, you are encouraged to use the loops
+
+#### Constant Variables
+
+A constant variable is an immutable variable that can be fully evaluated at parse-time. These
+are useful with commands that need to know the value of an argument at parse time,
+like [`source`](/commands/docs/source.md), [`use`](/commands/docs/use.md) and [`register`](/commands/docs/register.md). See [how nushell code gets run](how_nushell_code_gets_run.md)
+for a deeper explanation. They are declared using the `const` keyword
+
+```
+const plugin = 'path/to/plugin'
+register $plugin
 ```
-> let my_value = 4
-> $my_value
-4
+
+### Variable Names
+
+Variable names in nushell come with a few restrictions as to what characters they
+can contain. In particular, they cannot contain these characters
+
+```
+.  [  (  {  +  -  *  ^  /  =  !  <  >  &  |
+```
+
+It is common for some scripts to declare variables that start with `$`. This is
+allowed.
+
+```nu
+> let $var = 42
+# identical to `let var = 42`
 ```
 
-Variables in Nushell are immutable, that means that you cannot change its value after declaration.
-They can be shadowed in nested block, that results in:
+### Variable Values
+
+A common issue that users run into is trying to declare a variable using a pipeline
+as the value like so:
 
+```nu
+let val = 42 | math sin
 ```
-> let my_value = 4
-> do { let my_value = 5; $my_value }
-5
-> $my_value
-4
+
+This is an error, because the pipe command is used to separate pipeline elements, so
+nushell will see this as a pipeline with two elements, one of which is not allowed
+in pipleines because it doesn't return a value (`let val = 42`). The correct way to
+declare using a pipeline is to wrap the pipeline in parentheses
+
+```nu
+let val = (42 | math sin)
 ```
 
-## Variable paths
+### Variable Paths
```

Comment:
Hmm. I hadn't seen the Variable Paths section, I think we should remove it eventually.

[Cell paths](https://github.com/nushell/nushell.github.io/pull/879) aren't really something to do with a _variable_. The cell path `.bar` in `$foo.bar` is the same as the cell path in `(some-command).bar`.

That said, moving/removing this section is probably beyond the scope of this PR; we can do that another time. Thank you for documenting the optional `?` operator!